File: chapter-2/insertion-sort.rb
#!/usr/bin/env ruby

# Sort by Insertion Sort
#
# Try to imagine you get a things
# move bigger things to right
# then put it to the right place
module InsertionSort
  class << self
    def insertion_sort!(array)
      (1...array.length).each do |holding_index|
        holdind = get(array, holding_index)
        place_to = right_shift(array, holdind, holding_index)
        put(array, place_to, holdind)
      end
    end

    def insertion_sort(array)
      new = array.dup
      insertion_sort!(new)
      new
    end

    private_class_method

    def get(array, index)
      array[index]
    end

    def right_shift(array, holding, holding_index)
      right_shifting_index = holding_index - 1
      loop do
        break if right_shifting_index < 0
        need_right_shift = array[right_shifting_index]
        break if need_right_shift < holding # Get to the right place
        array[right_shifting_index + 1] = need_right_shift
        right_shifting_index -= 1
      end
      right_shifting_index + 1 # Correct - 1
    end

    def put(array, index, value)
      array[index] = value
    end
  end
end

# rubocop:disable MethodLength
def insertion_sort!(array)
  (1...array.length).each do |current_index|
    current = array[current_index]
    before_index = current_index - 1
    loop do
      break if before_index < 0
      before = array[before_index]
      break if before < current
      array[before_index + 1] = before
      before_index -= 1
    end
    array[before_index + 1] = current
  end
end
# rubocop:enable MethodLength

if $PROGRAM_NAME == __FILE__
  array = gets.chomp.split(' ').map(&:to_i)
  puts 'readable code:'
  puts InsertionSort.insertion_sort(array).join(' ')
  puts 'concise code:'
  insertion_sort!(array)
  puts array.join(' ')
end

File: chapter-2/insertion_sort.py
#!/usr/bin/env python

class InsertionSort:
    '''Sort by Insertion Sort

    Try to imagine you get a things
    move bigger things to right
    then put it to the right place
    '''
    def insertion_sort(array):
        for holding_index in range(1, len(array)):
            holdind = InsertionSort.__get(array, holding_index)
            place_to = InsertionSort.__right_shift(array, holdind, holding_index)
            InsertionSort.__put(array, place_to, holdind)

    def __get(array, index):
        return array[index]

    def __right_shift(array, holding, holding_index):
        right_shifting_index = holding_index - 1
        while True:
            if right_shifting_index < 0: break
            need_right_shift = array[right_shifting_index]
            if need_right_shift < holding: break # Get to the right place
            array[right_shifting_index + 1] = need_right_shift
            right_shifting_index -= 1
        return right_shifting_index + 1 # Correct - 1

    def __put(array, index, value):
        array[index] = value

def insertion_sort(array):
    for current_index in range(1, len(array)):
        current = array[current_index]
        before_index = current_index - 1
        while True:
            if before_index < 0: break
            before = array[before_index]
            if before < current: break
            array[before_index + 1] = before
            before_index -= 1
        array[before_index + 1] = current

if __name__ == '__main__':
    array1 = list(map(int, input().strip().split(' ')))
    array2 = array1[:]
    print('readable code:')
    InsertionSort.insertion_sort(array1)
    print(' '.join((str(x) for x in array1)))
    print('concise code:')
    insertion_sort(array2)
    print(' '.join((str(x) for x in array2)))
